fem/particleset.cpp

```diff
@@ -0,0 +1,902 @@
+   for (int f = 0; f < GetNFields(); f++)
+   {
+      field_vdims[f] = Field(f).GetVDim();
+   }
+   return field_vdims;
+}
+
+void ParticleSet::AddParticles(const Array<unsigned int> &new_ids,
+                               Array<int> *new_indices)
+{
+   int num_add = new_ids.Size();
+   int old_np = GetNParticles();
+   int new_np = old_np + num_add;
+
+   // Set indices of new particles
+   if (new_indices)
+   {
+      new_indices->SetSize(num_add);
+      for (int i = 0; i < num_add; i++)
+      {
+         (*new_indices)[i] = ids.Size() + i;
+      }
+   }
+   // Add new ids
+   ids.Append(new_ids);
+
+   // Update data
+   for (int f = -1; f < GetNFields(); f++)
+   {
+      ParticleVector &pv = (f == -1 ? coords : *fields[f]);
+      pv.SetNumParticles(new_np); // does not delete existing data
+   }
+
+   // Update tags
+   for (int t = 0; t < GetNTags(); t++)
+   {
+      tags[t]->SetSize(new_np);
+   }
+}
+
+#if defined(MFEM_USE_MPI) && defined(MFEM_USE_GSLIB)
+
+/// \cond DO_NOT_DOCUMENT
+template<std::size_t NBytes>
+void ParticleSet::TransferParticlesImpl(ParticleSet &pset,
+                                        const Array<unsigned int> &send_idxs, const Array<unsigned int> &send_ranks)
+{
+   struct pdata_t
+   {
+      alignas(double) std::array<std::byte, NBytes> data;
+      unsigned int id;
+   };
+
+   int nreals = pset.GetFieldVDims().Sum() + pset.Coords().GetVDim();
+   int ntags = pset.GetNTags();
+   // in-case particles have not been initialized on all ranks
+   MPI_Allreduce(MPI_IN_PLACE, &nreals, 1, MPI_INT, MPI_MAX, pset.GetComm());
+   MPI_Allreduce(MPI_IN_PLACE, &ntags, 1, MPI_INT, MPI_MAX, pset.GetComm());
+   int nbytes = nreals*sizeof(double) + ntags*sizeof(int);
+
+   using parr_t = pdata_t;
+   MFEM_VERIFY(nbytes <= NBytes, "More data than can be packed.");
+
+   gslib::array gsl_arr;
+   parr_t *pdata_arr;
+   array_init(parr_t, &gsl_arr, send_idxs.Size());
+   pdata_arr = (parr_t*) gsl_arr.ptr;
+
+   gsl_arr.n = send_idxs.Size();
+   for (int i = 0; i < send_idxs.Size(); i++)
+   {
+      parr_t &pdata = pdata_arr[i];
+      pdata.id = pset.GetIDs()[send_idxs[i]];
+
+      // Copy particle data directly into pdata
+      int counter = 0;
+      for (int f = -1; f < pset.GetNFields(); f++)
+      {
+         ParticleVector &pv = (f == -1 ? pset.Coords() : pset.Field(f));
+         for (int c = 0; c < pv.GetVDim(); c++)
+         {
+            std::memcpy(pdata.data.data() + counter, &pv(send_idxs[i], c),
+                        sizeof(double));
+            counter += sizeof(double);
+         }
+      }
+
+      // Copy tags
+      for (int t = 0; t < pset.GetNTags(); t++)
+      {
+         Array<int> &tag_arr = pset.Tag(t);
+         std::memcpy(pdata.data.data() + counter, &tag_arr[send_idxs[i]],
+                     sizeof(int));
+         counter += sizeof(int);
+      }
+   }
+
+   // Remove particles that will be transferred
+   pset.RemoveParticles(send_idxs);
```

Comment:
This is very inefficient 😮  as the data are shifted even when the number of received particles is the same as sent. The algorithm should be to receive particles first and place them at the "recycled" sending indices. If the size is not enough, allocate some extra. If it is less then only shift the data (ideally sorting the indices to remove the ones near the end of the vector for a minimal shift). That should be eliminated in the future as well by some flags for the particle state, I believe, but that is another matter for later.